pkg/chunked/bloom_filter.go

```diff
@@ -0,0 +1,80 @@
+package chunked
+
+import (
+	"encoding/binary"
+	"hash/crc32"
+	"io"
+)
+
+type bloomFilter struct {
+	bitArray []uint64
+	k        uint32
+}
+
+func newBloomFilter(size int, k uint32) *bloomFilter {
+	numElements := (size + 63) / 64
+	return &bloomFilter{
+		bitArray: make([]uint64, numElements),
+		k:        k,
+	}
+}
+
+func newBloomFilterFromArray(bitArray []uint64, k uint32) *bloomFilter {
+	return &bloomFilter{
+		bitArray: bitArray,
+		k:        k,
+	}
+}
+
+func (bf *bloomFilter) hashFn(item []byte, seed uint32) (uint64, uint64) {
+	mod := uint32(len(bf.bitArray) * 64)
+	hash := (seed ^ crc32.ChecksumIEEE(item)) % mod
```

Comment:
Is the way `seed` is integrated here sufficient? Apart from the `% mod` part, this seems to be effectively computing 3 hashes of (crc32, crc32 ^ 1, crc32 ^ 2); intuitively (I didn’t do the math) 3 hash functions like this should be about as effective as using a single hash — a collision in the first hash implies a collision in the other hashes.

For the same reason, doing a CRC32 step “appending” `seed` to `ChecksumIEEE(item)` wouldn’t work well, and AFAICS integrating the seed as a data input to CRC32 wouldn’t work well either (all inputs are of the same length, so the seed affect the same 2 bits of output).

Ugh… I didn’t expect to be evaluating independence of hash functions this week ;)

---

It’s quite possible this is a big enough improvement with CRC32… we can always bump the cache version later to change the functions.

Reply:
good point.

I've played more with it and use the seed to split item in two slices and checksum each of them.  The benchmark tests still show a benefit for using the bloom filter (with 3 hash functions):

```
$ go test -bench=.
goos: linux
goarch: amd64
pkg: github.com/containers/storage/pkg/chunked
cpu: Intel(R) Core(TM) i7-10850H CPU @ 2.70GHz
BenchmarkLookupBloomFilter-12    	1000000000	         0.08671 ns/op
BenchmarkLookupBloomRaw-12       	1000000000	         0.1495 ns/op
PASS
ok  	github.com/containers/storage/pkg/chunked	2.628s
```

Reply:
OK. I can’t immediately see that this is not uniform, so I’ll go with the practical evidence :)